Consolidate L4 ports for an ACL

venuiyer opened the pull request:
Currently, if we have a k8s nw policy such as, e.g :

  - from:
    - namespaceSelector:
          ...
      podSelector:
          ...
    ports:
    - protocol: TCP
      port: 80
    - protocol: TCP
      port: 3000
    - protocol: TCP
      port: 8080

We'll have logical flows that look like:
  to-lport  1001 (ip4.src == {$as1} && tcp && tcp.dst==3000 && outport == @pg1) allow-related
  to-lport  1001 (ip4.src == {$as1} && tcp && tcp.dst==80 && outport == @pg1) allow-related
  to-lport  1001 (ip4.src == {$as1} && tcp && tcp.dst==8080 && outport == @pg1) allow-related

It'll be better to consolidate this as:
  to-lport  1001 (ip4.src == {$as1} && tcp && tcp.dst=={80,3000,8080} && outport == @pg1) allow-related

This allows OVN controller to reduce the number of flows using conjunction rules.
Did functional testing for single and multiple port TCP/UDP network policies.

Reported-by: Han Zhou <hzhou@ovn.org>
Signed-off-by: venu iyer <venugopali@nvidia.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/ovn-org/ovn-kubernetes/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Trivial changes are exempt from following this template.
If your change is non-trivial, please provide the following information:
-->

**- What this PR does and why is it needed**
<!--
A summary of the changes within this pull request and some context
as to why they were made
-->

**- Special notes for reviewers**
<!--
What exactly did you change - you may also defer to information
contained in commit messages. At a bare minimum it's worth highlighting
which areas of the code were changed as it's easier to assign reviewers
-->


**- How to verify it**
<!--
Did you include unit tests? or end-to-end tests?
How can I manually verify that this patch achieves its objective
-->


**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

girishmg commented:
@venuiyer how are you handling upgrades? What will happen to all the existing ACLs that have just one port? If you are not handling upgrades, then you should. Please look at the pkg/ovn/master.go

venuiyer commented:
> @venuiyer how are you handling upgrades? What will happen to all the existing ACLs that have just one port? If you are not handling upgrades, then you should. Please look at the pkg/ovn/master.go

Thanks, @girishmg.  Do you mean update all the existing acl  using this? i was not thinking about that; understand that on a large scale with existing acl, that might be helpful, but i was not sure about the  added complexity .. since we'll need to walk all the pg/acl and check for modification; that'll also cause a  churn on the logical and controller side during the upgrade.

girishmg commented:
> > @venuiyer how are you handling upgrades? What will happen to all the existing ACLs that have just one port? If you are not handling upgrades, then you should. Please look at the pkg/ovn/master.go
> 
> Thanks, @girishmg. Do you mean update all the existing acl using this? i was not thinking about that; understand that on a large scale with existing acl, that might be helpful, but i was not sure about the added complexity .. since we'll need to walk all the pg/acl and check for modification; that'll also cause a churn on the logical and controller side during the upgrade.

yes, the churn will be one time hit during the upgrade. however, after the upgrade churn the benefit because of combined ACLs should help, right? Otherwise, without the upgrade path there is no benefit for an already running k8s clusters and static network policies

also, i believe there is code in place which check for `old` match string with the newly generated match string and then modify the ACL only if they have changed. with your changes, that code will always fail since in old method each port is an ACL and nin new method all the ports are combined into one ACL.

venuiyer commented:
> > > @venuiyer how are you handling upgrades? What will happen to all the existing ACLs that have just one port? If you are not handling upgrades, then you should. Please look at the pkg/ovn/master.go
> > 
> > 
> > Thanks, @girishmg. Do you mean update all the existing acl using this? i was not thinking about that; understand that on a large scale with existing acl, that might be helpful, but i was not sure about the added complexity .. since we'll need to walk all the pg/acl and check for modification; that'll also cause a churn on the logical and controller side during the upgrade.
> 
> yes, the churn will be one time hit during the upgrade. however, after the upgrade churn the benefit because of combined ACLs should help, right? Otherwise, without the upgrade path there is no benefit for an already running k8s clusters and static network policies
> 
> also, i believe there is code in place which check for `old` match string with the newly generated match string and then modify the ACL only if they have changed. with your changes, that code will always fail since in old method each port is an ACL and nin new method all the ports are combined into one ACL.



> > > @venuiyer how are you handling upgrades? What will happen to all the existing ACLs that have just one port? If you are not handling upgrades, then you should. Please look at the pkg/ovn/master.go
> > 
> > 
> > Thanks, @girishmg. Do you mean update all the existing acl using this? i was not thinking about that; understand that on a large scale with existing acl, that might be helpful, but i was not sure about the added complexity .. since we'll need to walk all the pg/acl and check for modification; that'll also cause a churn on the logical and controller side during the upgrade.
> 
> yes, the churn will be one time hit during the upgrade. however, after the upgrade churn the benefit because of combined ACLs should help, right? Otherwise, without the upgrade path there is no benefit for an already running k8s clusters and static network policies
> 
> also, i believe there is code in place which check for `old` match string with the newly generated match string and then modify the ACL only if they have changed. with your changes, that code will always fail since in old method each port is an ACL and nin new method all the ports are combined into one ACL.

Sorry for the delay, @girishmg Thanks for pointing that out - else on restart we'll be left with redundant flows. Updated the PR with the change, please take a look when you get a chance; tested with ipblock and multiple tcp flows to check. Thanks!

hzhou8 commented on go-controller/pkg/ovn/gress_policy.go:
```diff
@@ -45,24 +45,54 @@ type portPolicy struct {
 }
 
-func (pp *portPolicy) getL4Match() (string, error) {
+func (pp *portPolicy) getL4Match(protoFilters map[string]string) error {
```
nit: naming of this function is now a little confusing. It is now appending new matches instead of just "get". So I suggest to rename the function to appendL4Match. In addition, the param protoFilters is better to match the naming of the function, e.g. l4Matches.

hzhou8 commented on go-controller/pkg/ovn/gress_policy.go:
```diff
@@ -45,24 +45,54 @@ type portPolicy struct {
+		// If we already permit all TCP traffic, we don't need to add the ports.
+		// If there is an update that removes blanket tcp, then we'll reevaluate the
+		// matches. Same for the other protocols too
+		if protoFilters["tcp"] == "tcp" {
```
if xxx["tcp"] == "tcp" looks confusing. Since the value is supposed to be L4 ports, and the keyword here is supposed to mean "all ports", it would be more clear to just use "all" as the keyword. With the keyword "all", this function can be very simple because the 3 branches for different protocols can be combined to one common logic.

hzhou8 reviewed (commented):
Thanks Venu for the optimization. Please see my comments.

hzhou8 commented on go-controller/pkg/ovn/gress_policy.go:
```diff
@@ -59,15 +59,35 @@ func (pp *portPolicy) getL4Match() (string, error) {
+			l4Matches[foundProtocol]["portrange"] = fmt.Sprintf("%d<=%s.dst<=%d", pp.port,
+				foundProtocol, pp.endPort)
+		} else {
+			l4Matches[foundProtocol]["portrange"] = fmt.Sprintf("%s || %d<=%s.dst<=%d",
```
If there are overlapping between port ranges, are they merged already before calling this function?

hzhou8 replied:
Sounds good. It can be a separate PR (if needed).

hzhou8 replied:
Yes, LGTM. Thanks @venuiyer

hzhou8 commented on go-controller/pkg/ovn/policy_test.go:
```diff
@@ -1095,14 +1097,32 @@ var _ = ginkgo.Describe("OVN NetworkPolicy Operations", func() {
 							Protocol: &tcpProtocol,
-						}},
+						},
+							{
```
nit: indentation problem

girishmg commented:
@venuiyer Sorry, I should have reviewed this PR sooner.

So, your upgrade logic is not a one time thing. Everytime a network policy is added we try to craft the old way of authoring ACLs and try to delete it. So, if this PR is merged, then any new installation will always see an addition of New ACLs followed by an attempt to delete the OLD ACLs. This will be true forever.

All our upgrades in the past has been one time thing that we do when we start the ovnkube-master... for example upgradeToSingleSwitchOVNTopology() and we increment our logical topology version... subsequent boots we don't do the upgrade anymore.

Not sure what others think? @trozet @dbw for viz

astoycos commented on go-controller/pkg/ovn/libovsdbops/portgroup.go:
```diff
@@ -271,10 +271,10 @@ func DeleteACLsFromPortGroupOps(nbClient libovsdbclient.Client, ops []libovsdb.O
 		err := findACL(nbClient, acl)
-		if err != nil {
-			return nil, err
+		// Proceed to check others if we don't find an ACL.
```
I _think_ this is wrong, we still wan't to check the err, however if you wan't to keep iterating even after failing to find an ACL (i.e  `errNotFound` error) You can do 

```if err !=nil && err != libovsdbclient.ErrNotFound```

venuiyer commented:
> This is kind of a funny upgrade scenario... Since we're changing the number of ACLs not just mutating the existing ACLs (which is much easier now with libovsdb see [CreateOrUpdate](https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/libovsdbops/model_client.go#L125))
> 
> > All our upgrades in the past has been one time thing that we do when we start the ovnkube-master... for example upgradeToSingleSwitchOVNTopology() and we increment our logical topology version... subsequent boots we don't do the upgrade anymore.
> 
> Yep, usually we take care of existing stale resources in our "sync" functions at startup
> 
> > I didn't think of versioning Network Policy as i am not sure if that'll be an overkill for this purpose.
> 
> I don't think versioning would be the right route here, for now lets just take care of stale acls in a one time op as noted above.
> 
> I think to also future proof this it'd be smart to add an external ID to each allow ACL to correspond it to it's parent network policy (like `netPol=<netpol name>`), then we would easily be able to find all ACLs for a given policy and update them on an as needed basis (rather than having to calculate old matches for lookups since now the match conditions have changed) as the ACL implementation continues to evolve.

Sorry for the delay in response @astoycos / @girishmg . I suppose i can move the deletion to syncNetworkPolicies (and differentiate stale entries using, say, an external id), but unless i am reading it wrong, we'll clean up only on a subsequent restart, right? If so,  after upgrading to the ovn-k8s with this change, till we restart the next time, we'll have redundant flows (which defeats the purpose of this change, maybe in the short run); is that acceptable? (btw, w.r.t to 'netpol=<netpol name>' comment; the network policy name is in the external ids for the acl currently? -- e.g. "external_ids        : {... namespace=<ns>, policy=<policyname>, ...}"). let me know if i am missing something. Thanks!

venuiyer commented:
@girishmg @astoycos a happy new year! sorry for the delay, just got back to this. Please take a look at the revised changes and let me know if it addresses the concerns. I added a version identifier, but am leveraging  cleanup after bringing up everything instead of  converting upfront,  as I think it is better to keep it simple rather than try to optimize this one time operation. I did distinguish the new acl at the expense of adding an ext id to NB  as i was a  hesitant to rename the existing ext id (to differentiate) since i was not sure if any application uses it. Please have a look and let me know your comments etc. Thanks very much!

tssurya commented on go-controller/pkg/types/const.go:
```diff
@@ -108,7 +108,8 @@ const (
 	OvnRoutingViaHostTopoVersion   = 5
-	OvnCurrentTopologyVersion      = OvnRoutingViaHostTopoVersion
+	OvnL4ACLConsolidatedVersion    = 6
+	OvnCurrentTopologyVersion      = OvnL4ACLConsolidatedVersion
```
based on upstream meeting discussion: let's remove the topology version bump and use the same logic as egress firewall.

dcbw commented:
Obsoleted by https://github.com/ovn-org/ovn-kubernetes/pull/2950
